### ISSUE TITLE
publish and pack commands have confusing path expectations

### ISSUE THREAD

**Issue by ashleygwilliams** [rustwasm/wasm-pack#189]
currently the `publish` and `pack` commands anticipate that you are running them from a directory that contains a `pkg` directory to be published. i have confusingly `cd`'d into the `pkg` directory and tried and gotten an error, e.g.:

```
Ashleys-MBP-2:wasm-pack ag_dubs$ cd tests/fixtures/js-hello-world/pkg/
Ashleys-MBP-2:pkg ag_dubs$ wasm-pack publish
| ⛔   [ERR]: There was an I/O error. Details:
No such file or directory (os error 2)
Caused by: No such file or directory (os error 2)
```

i think it's working fine as is, but perhaps we want to have a custom error message here that explains the situation, and/or have an option to pass a path to the pkg directory, not just the parent directory? unclear exactly what's the best UX here- would love to hear ya'lls thoughts!

**Comment by Mackiovello** [comment]
I would like to give this a try! I think it should work similarly to how [`dotnet publish`](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-publish?tabs=netcore21#arguments) works: 

* It uses the current directory by default
* You can optionally set the directory as the first argument
* If there's nothing to publish in the directory, it gives a nice error message: `Specify a project or solution file. The current working directory does not contain a project or solution file.`

It may not be the best comparison, but it's very similar to what you propose and what feels intuitive in my mind. 

I'm relatively new to Rust, but I think I can implement this and create a pull request.

**Comment by mgattozzi** [comment]
Yeah! I believe this is the correct way and kind of what we would want to do. So what happens is that it just uses the I/O error which is "Can't find the file". Best way to make this nicer is the following:

- Before we publish create a `Path` to the `package.json`
- Check that it [exists](https://doc.rust-lang.org/std/path/struct.Path.html#method.exists)
- If it does not keep trying the different methods you mentioned for path resolution
- If it exists then continue with the publishing
- Otherwise throw an error

The key thing here is you'll need to create a new error so that you can have that custom string you mentioned! You can do that in `src/error.rs`

Thanks for taking this! Feel free to ping here or in a WIP PR if you need help :D

**Comment by Mackiovello** [comment]
After looking at this some more, I think it's surprising that the user-entered path should point to the parent of the pkg directory. I don't know any tool that works this way. Would it be alright to change this behavior? 

What I imagine is that the user-entered path points directly to the pkg directory by default. But to maintain backwards compatibility, it also looks in the child directories if the entered path is not the pkg directory.

The default path should use the current directory by default, and it should be able to find the pkg directory no matter if it's in the child directory of the current directory or if it's pointing to the pkg directory.

> Before we publish create a Path to the package.json

I don't know what `package.json` has to do with this. Both the `pack` and `publish` command only need the pkg directory. Am I missing something here? 

As a side note, when I use clippy on this repo, I get a bunch of warning, but that's for another issue.

**Comment by mgattozzi** [comment]
@ashleygwilliams was this closed by #198 or was that only closing part of the issue? I'm catching up on two weeks of things here so I'm a bit uncertain as to that.

**Comment by ashleygwilliams** [comment]
closed by #198
